### PR TITLE
Reuse AssemblyName objects to reduce memory pressure

### PR DESCRIPTION
**Pull request by Therzok**
Fixes https://github.com/microsoft/vs-mef/issues/251

### CODE REVIEW

File: src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs [hunk 1]
```diff
@@ -65,7 +65,7 @@ public class TypeRef : IEquatable<TypeRef>, IEquatable<Type>
             Requires.NotNullOrEmpty(fullName, nameof(fullName));
 
             this.resolver = resolver;
-            this.AssemblyName = GetNormalizedAssemblyName(assemblyName);
+            this.AssemblyName = Resolver.GetNormalizedAssemblyName(assemblyName);
```
Comment (Therzok) [inline src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs:68]: These won't use cached versions

**Review by AArnott — APPROVED** [review]
I like it. Thank you very much for submitting.
Does your local testing show that this substantially reduces the problem you set out to fix?

File: src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs [hunk 2]
```diff
@@ -238,7 +238,7 @@ internal Type ResolvedType
         public static TypeRef Get(Resolver resolver, AssemblyName assemblyName, int metadataToken, string fullName, TypeRefFlags typeFlags, int genericTypeParameterCount, ImmutableArray<TypeRef> genericTypeArguments, bool shallow, ImmutableArray<TypeRef> baseTypes, TypeRef? elementTypeRef)
         {
             Requires.NotNull(resolver, nameof(resolver));
-            return new TypeRef(resolver, assemblyName, null, metadataToken, fullName, typeFlags, genericTypeParameterCount, genericTypeArguments, shallow, baseTypes, elementTypeRef);
+            return new TypeRef(resolver, resolver.GetNormalizedAssemblyName(assemblyName), null, metadataToken, fullName, typeFlags, genericTypeParameterCount, genericTypeArguments, shallow, baseTypes, elementTypeRef);
```
Comment (AArnott) [inline src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs:241]: Build break would be fixed by this, I think:

```suggestion
            return new TypeRef(resolver, Resolver.GetNormalizedAssemblyName(assemblyName), null, metadataToken, fullName, typeFlags, genericTypeParameterCount, genericTypeArguments, shallow, baseTypes, elementTypeRef);
```

Comment (Therzok) [reply src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs:241]: Yeah, I had uncommitted changes in my local checkout.

I will add tests for the PR aswell

**Comment by Therzok** [comment]
Just tested this PR locally, down to 152 AssemblyName instances (19kb)

File: src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs (hunk 1 above)
Comment (Therzok) [reply src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs:68]: Sounds good, thank you! I scanned the caller tree a bit and seemed as much, but wanted to get confirmation. Resolving!

File: test/Microsoft.VisualStudio.Composition.Tests/Reflection/TypeRefTests.cs [hunk 3]
```diff
@@ -21,6 +21,15 @@ public void EqualsDistinguishesArrays()
             Assert.NotEqual(TypeRef.Get(typeof(object), TestUtilities.Resolver), TypeRef.Get(typeof(object[]), TestUtilities.Resolver));
         }
 
+        [Fact]
+        public void AssemblyNameInstancesReused()
+        {
+            var typeRef1 = TypeRef.Get(typeof(object), TestUtilities.Resolver);
```
Comment (Therzok) [inline test/Microsoft.VisualStudio.Composition.Tests/Reflection/TypeRefTests.cs:27]: Added a really simple test to ensure this works and doesn't regress out of the blue.

**Comment by Therzok** [comment]
@AArnott PTAL

**Review by AArnott — APPROVED** [review]

**Comment by AArnott** [comment]
@therzok Thanks again!

**Pull request merged**